File: .github/scripts/validate-cli-structured-output-contract.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

SKILL_PATH = "SKILL.md"
CLI_PATH = "CLI_INTERFACE.md"

unless File.file?(SKILL_PATH)
  warn "Missing universally required file: #{SKILL_PATH}"
  exit 1
end

strip_backticks = lambda do |value|
  normalized = value.to_s.strip
  if normalized.length >= 2 && normalized.start_with?("`") && normalized.end_with?("`")
    normalized[1...-1]
  else
    normalized
  end
end

markdown_section = lambda do |document, heading|
  level = heading[/\A#+/].length
  boundary = level == 2 ? "^##\\s|\\z" : "^(?:##|###)\\s|\\z"
  match = document.match(
    Regexp.new("^#{Regexp.escape(heading)}\\s*$\\n(.*?)(?=#{boundary})", Regexp::MULTILINE)
  )
  match && match[1]
end

field_value = lambda do |section, label|
  match = section&.match(/^#{Regexp.escape(label)}:\s*(.*?)\s*$/)
  match && strip_backticks.call(match[1])
end

profile_values = File.readlines(SKILL_PATH, chomp: true).filter_map do |raw_line|
  normalized = raw_line.strip
  normalized = normalized[2..].strip if normalized.start_with?("- ")
  match = normalized.match(/\ASelected profiles:\s*(.+?)\s*\z/)
  strip_backticks.call(match[1]) if match
end

if profile_values.length != 1
  warn "#{SKILL_PATH} must contain exactly one 'Selected profiles:' declaration."
  exit 1
end

selected_profiles = profile_values.first.split(",").map(&:strip).reject(&:empty?)
if selected_profiles == ["template-scaffold"] || !selected_profiles.include?("packaged-cli")
  puts "CLI structured-output contract is not activated."
  exit 0
end

errors = []

unless File.file?(CLI_PATH)
  errors << "Selected profile 'packaged-cli' requires contract file: #{CLI_PATH}"
else
  cli = File.read(CLI_PATH)
  structured = markdown_section.call(cli, "### Structured output")

  if structured.nil? || structured.strip.empty?
    errors << "#{CLI_PATH} requires a non-empty '### Structured output' section."
  else
    mode_selector = field_value.call(structured, "Mode selector")
    format = field_value.call(structured, "Format")
    version_field = field_value.call(structured, "Contract version field")

    unresolved_selector = /\A(?:NONE|NOT\s+(?:SUPPORTED|APPLICABLE)|TODO|TBD|FIXME|PLACEHOLDER|UNSELECTED|PENDING|AUTOMATIC|DEFAULT|SEE\s+DOCUMENTATION)\z/i
    unresolved_selector_payload = /(?:\A|[:=]|\s)(?:NONE|NOT\s+(?:SUPPORTED|APPLICABLE)|TODO|TBD|FIXME|PLACEHOLDER|UNSELECTED|PENDING|AUTOMATIC|DEFAULT|SEE\s+DOCUMENTATION)(?=\z|[\s.,;])/i
    option_selector = /(?:\A|\s)--?[A-Za-z0-9][A-Za-z0-9_-]*(?:[=\s]\S+)?(?:\s|\z)/
    environment_selector = /(?:\A|\s)[A-Z][A-Z0-9_]*=\S+(?:\s|\z)/
    named_selector = /\A(?:subcommand|command|option|flag)\s*:\s*\S(?:.*\S)?\z/i
    selector_is_explicit = mode_selector && (
      option_selector.match?(mode_selector) ||
      environment_selector.match?(mode_selector) ||
      named_selector.match?(mode_selector)
    )
    selector_is_resolved = mode_selector &&
                           !unresolved_selector.match?(mode_selector) &&
                           !unresolved_selector_payload.match?(mode_selector)
    unless selector_is_explicit && selector_is_resolved
      errors << "#{CLI_PATH} 'Mode selector:' must record an exact, fully resolved caller-visible option, subcommand, or environment assignment that activates structured output."
    end

    # Format names are intentionally open-ended. A concrete skill may select a
    # standard or application-specific serialization without waiting for this
    # template's validator to add it to a whitelist. Reject only declarations
    # that explicitly deny structured machine readability, generic nonchoices,
    # placeholders, or prose too vague to identify a serialization contract.
    negative_format = /\b(?:PLAIN\s+TEXT|HUMAN[-\s]+READABLE|TEXT\s+ONLY|NO\s+STRUCTURED|UNSTRUCTURED|NOT\s+MACHINE[-\s]+READABLE)\b/i
    generic_format = /\A(?:TEXT|BINARY|CUSTOM|OTHER|UNKNOWN|NONE|NOT\s+(?:SUPPORTED|APPLICABLE)|TODO|TBD|UNSELECTED)\z/i
    format_name = /\A(?=.{1,80}\z)(?:[A-Za-z0-9][A-Za-z0-9._+\/-]*)(?:[ -][A-Za-z0-9][A-Za-z0-9._+\/-]*){0,5}\z/
    unless format && format_name.match?(format) && !negative_format.match?(format) && !generic_format.match?(format)
      errors << "#{CLI_PATH} 'Format:' must name an explicit machine-readable structured serialization format."
    end

    negative_version = /\A(?:NONE|NOT\s+(?:SUPPORTED|APPLICABLE)|NO\s+VERSION\s+FIELD)\z|\b(?:WITHOUT|OMITTED|ABSENT)\b/i
    field_selector = /\A(?:[A-Za-z_$][A-Za-z0-9_$-]*(?:\.[A-Za-z_$][A-Za-z0-9_$-]*)*|\/(?:[^\/\s]+\/)*[^\/\s]+)\z/
    unless version_field && !negative_version.match?(version_field) && field_selector.match?(version_field)
      errors << "#{CLI_PATH} 'Contract version field:' must name one concrete field or field path."
    end
  end
end

unless errors.empty?
  errors.uniq.each { |error| warn error }
  exit 1
end

puts "CLI structured-output contract is valid."

File: .github/scripts/validate_cli_structured_output_contract.py
#!/usr/bin/env python3
"""Validate the packaged CLI structured-output contract."""

from __future__ import annotations

import re
import sys
from pathlib import Path

from lib.profile_contracts import MarkdownDocument, ParseError, ProfileSelection


SKILL_PATH = Path("SKILL.md")
CLI_PATH = Path("CLI_INTERFACE.md")


def run() -> int:
    try:
        selection = ProfileSelection.load(SKILL_PATH)
    except (ParseError, OSError) as exc:
        print(exc, file=sys.stderr)
        return 1

    if selection.template_scaffold() or not selection.selected("packaged-cli"):
        print("CLI structured-output contract is not activated.")
        return 0

    errors: list[str] = []
    if not CLI_PATH.is_file():
        errors.append(
            f"Selected profile 'packaged-cli' requires contract file: {CLI_PATH}"
        )
    else:
        cli = MarkdownDocument.read(CLI_PATH)
        structured = cli.section("### Structured output")

        if structured is None or not structured.strip():
            errors.append(
                f"{CLI_PATH} requires a non-empty '### Structured output' section."
            )
        else:
            structured_document = MarkdownDocument(structured, path=CLI_PATH)
            mode_selector = structured_document.field("Mode selector")
            format_name_value = structured_document.field("Format")
            version_field = structured_document.field("Contract version field")

            unresolved_selector = re.compile(
                r"^(?:NONE|NOT\s+(?:SUPPORTED|APPLICABLE)|TODO|TBD|FIXME|"
                r"PLACEHOLDER|UNSELECTED|PENDING|AUTOMATIC|DEFAULT|"
                r"SEE\s+DOCUMENTATION)$",
                re.IGNORECASE,
            )
            unresolved_selector_payload = re.compile(
                r"(?:^|[:=]|\s)(?:NONE|NOT\s+(?:SUPPORTED|APPLICABLE)|TODO|TBD|"
                r"FIXME|PLACEHOLDER|UNSELECTED|PENDING|AUTOMATIC|DEFAULT|"
                r"SEE\s+DOCUMENTATION)(?=$|[\s.,;])",
                re.IGNORECASE,
            )
            option_selector = re.compile(
                r"(?:^|\s)--?[A-Za-z0-9][A-Za-z0-9_-]*"
                r"(?:[=\s]\S+)?(?:\s|$)"
            )
            environment_selector = re.compile(
                r"(?:^|\s)[A-Z][A-Z0-9_]*=\S+(?:\s|$)"
            )
            named_selector = re.compile(
                r"^(?:subcommand|command|option|flag)\s*:\s*\S(?:.*\S)?$",
                re.IGNORECASE,
            )

            selector_is_explicit = bool(
                mode_selector
                and (
                    option_selector.search(mode_selector)
                    or environment_selector.search(mode_selector)
                    or named_selector.fullmatch(mode_selector)
                )
            )
            selector_is_resolved = bool(
                mode_selector
                and not unresolved_selector.fullmatch(mode_selector)
                and not unresolved_selector_payload.search(mode_selector)
            )
            if not selector_is_explicit or not selector_is_resolved:
                errors.append(
                    f"{CLI_PATH} 'Mode selector:' must record an exact, fully "
                    "resolved caller-visible option, subcommand, or environment "
                    "assignment that activates structured output."
                )

            negative_format = re.compile(
                r"\b(?:PLAIN\s+TEXT|HUMAN[-\s]+READABLE|TEXT\s+ONLY|"
                r"NO\s+STRUCTURED|UNSTRUCTURED|"
                r"NOT\s+MACHINE[-\s]+READABLE)\b",
                re.IGNORECASE,
            )
            generic_format = re.compile(
                r"^(?:TEXT|BINARY|CUSTOM|OTHER|UNKNOWN|NONE|"
                r"NOT\s+(?:SUPPORTED|APPLICABLE)|TODO|TBD|UNSELECTED)$",
                re.IGNORECASE,
            )
            explicit_format_name = re.compile(
                r"^(?=.{1,80}$)(?:[A-Za-z0-9][A-Za-z0-9._+/-]*)"
                r"(?:[ -][A-Za-z0-9][A-Za-z0-9._+/-]*){0,5}$"
            )
            if not (
                format_name_value
                and explicit_format_name.fullmatch(format_name_value)
                and not negative_format.search(format_name_value)
                and not generic_format.fullmatch(format_name_value)
            ):
                errors.append(
                    f"{CLI_PATH} 'Format:' must name an explicit machine-readable "
                    "structured serialization format."
                )

            negative_version = re.compile(
                r"^(?:NONE|NOT\s+(?:SUPPORTED|APPLICABLE)|"
                r"NO\s+VERSION\s+FIELD)$|"
                r"\b(?:WITHOUT|OMITTED|ABSENT)\b",
                re.IGNORECASE,
            )
            field_selector = re.compile(
                r"^(?:[A-Za-z_$][A-Za-z0-9_$-]*"
                r"(?:\.[A-Za-z_$][A-Za-z0-9_$-]*)*|"
                r"/(?:[^/\s]+/)*[^/\s]+)$"
            )
            if not (
                version_field
                and not negative_version.search(version_field)
                and field_selector.fullmatch(version_field)
            ):
                errors.append(
                    f"{CLI_PATH} 'Contract version field:' must name one concrete "
                    "field or field path."
                )

    if errors:
        for error in dict.fromkeys(errors):
            print(error, file=sys.stderr)
        return 1

    print("CLI structured-output contract is valid.")
    return 0


if __name__ == "__main__":
    raise SystemExit(run())
